### ISSUE TITLE
Extension Doesn't Work with AutoMapper 6.2.0 In Integration Tests

### ISSUE THREAD

**Issue by adamrodger** [AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection#36]
When [integration testing with ASP.Net Core](https://docs.microsoft.com/en-us/aspnet/core/testing/integration-testing), it is very common to use `TestServer` to start your service as an in-memory host so that you can perform service-level integration tests via a real `HttpClient`.

However, since AutoMapper 6.2.0 was released this no longer works. Your `Startup` class is effectively called multiple times within the same process, so the `AddAutoMapper()` call happens multiple times, so the `Mapper.Initialize()` call also happens multiple times. This results in an exception in the second test:

```
System.InvalidOperationException : Mapper already initialized. You must call Initialize once per application domain/process.
   at AutoMapper.Mapper.set_Configuration(IConfigurationProvider value)
   at AutoMapper.Mapper.Initialize(Action`1 config)
   at AutoMapper.ServiceCollectionExtensions.AddAutoMapperClasses(IServiceCollection services, Action`1 additionalInitAction, IEnumerable`1 assemblies)
   at MyService.Startup.ConfigureServices(IServiceCollection services)
```

How can we 'clear out' the AutoMapper statics so that integration testing can be supported in a single process?

**Comment by lbargaoanu** [comment]
https://github.com/AutoMapper/AutoMapper/issues/2399

**Comment by jbogard** [comment]
I think we'll need a `Mapper.Reset` again because the built-in DI container for MS doesn't support configuring child containers (that's how I usually accomplish this).

**Comment by lbargaoanu** [comment]
I think a cleaner solution would be to make AddAutoMapperClasses receive a MapperConfiguration. Maybe use the static one by default.

**Comment by adamrodger** [comment]
From looking at the code, I agree with @lbargaoanu. I've made a local version of the extension which stops calling `Mapper.Initialize` and now all my integration tests pass. This is:

```csharp
public static IServiceCollection AddAutoMapper(IEnumerable<Assembly> assemblies)
{
    var profiles = assemblies.SelectMany(a => a.ExportedTypes).Where(baseType.IsAssignableFrom);

    // create a new config with all those profiles registered
    var config = new MapperConfiguration(cfg =>
    {
        foreach (Type profile in profiles)
        {
            cfg.AddProfile(profile);
        }
    });

    services.TryAddSingleton<IMapper>(sp => new Mapper(config, sp.GetService));

    return services;
}
```

**Comment by jbogard** [comment]
Adding a `MapperConfiguration` parameter would work, I think the problem is we do a ton of stuff in the constructor there. It'd probably need to be an `IMapperConfigurationExpression` instance instead?

That would help as well with people wanting to call `services.AddAutoMapper` multiple times.

**Comment by lbargaoanu** [comment]
Whatever works, I guess, but if you pass something, it gets used, otherwise, the same old. I think :)

**Comment by lbargaoanu** [comment]
@adamrodger Yes, it's safe.

**Comment by jbogard** [comment]
Ah, that won't work either because I still have to create `MapperConfiguration` and register it with the container. And that thing's immutable.

There was another PR that added a bit more work by doing an external mapper builder but it was quite complicated.

There's already so many overloads that I think I'd have to build an options object. And I don't feel like breaking so much right now :)

The easiest thing is probably `Mapper.Reset` and call it a day.

**Comment by adamrodger** [comment]
I dunno, I think my proof-of-concept works fine. It means you can't call AddAutoMapper() twice within the same Startup class, but I don't really see much of a use case for that. You can only register IMapper once regardless of the transient/scoped/singleton scope so it doesn't make sense to call it twice anyway.

That's a better solution than Mapper.Reset because you don't need to remember to call it in tear down code. You'll just get a fresh IMapper each time you start the service up in an integration test.

**Comment by jbogard** [comment]
Yours works but I can’t add that to the project as is.

**Comment by adamrodger** [comment]
I've raised a MR to show the potential solution. All the tests still pass and that works within my integration tests which previously were failing.

The difference is that the static `Mapper` is now not initialised so you can't do `Mapper.Map` from anywhere you want, but the whole point of registering with DI in the first place is that you want to inject the `IMapper` instead of using statics isn't it?

**Comment by lbargaoanu** [comment]
The mapper has to be scoped so it can receive per request dependencies.

**Comment by adamrodger** [comment]
Ah OK, so as long as the configuration is singleton (which is hopefully
where the heavy work happens) then there shouldn't be an issue with the
scope of the mapper itself..?

If that's the case, doesn't the mapper need to be transient so it can also
receive transient dependencies?

**Comment by lbargaoanu** [comment]
The mapper needs to create things per request, such as resolvers that receive a DbContext. The configuration needs to be a singleton. Other than that, I don't see anything wrong with your solution. But apparently @jbogard disagrees.

**Comment by adamrodger** [comment]
I've updated the MR and added tests to ensure the configuration is singleton but the mapper is scoped.

**Comment by erikaspl** [comment]
with AutoMapper 6.2.1 I was able to resolve Mapper already initialized issue by registering Mapper.Reset() action for application shutdown in my startup's configure method:

```        
public void Configure(IApplicationBuilder app, IHostingEnvironment env, 
            ILoggerFactory loggerFactory, IApplicationLifetime applicationLifetime)
    {
         .....
         applicationLifetime.ApplicationStopped.Register(() => AutoMapper.Mapper.Reset());
    }
```

When make sure TestServer dispose method is called in you test teardown:

```
        protected TestServer Server;
        [OneTimeSetUp]
        public void OneTimeBaseSetUp()
        {
             Server = new TestServer(....
        }


        [OneTimeTearDown]
        public void OneTimeBaseTearDown()
        {
            Server.Dispose();
        }

**Comment by mihaj** [comment]
I am also having the same issue. I use xUnit.

**Comment by sm-g** [comment]
@erikaspl So TestServer in your solution created only once for all tests? Not acceptable in some projects.
Adding (code from [docs](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/testing#integration-testing))
```cs
protected IntegrationTestFixture(string solutionRelativeTargetProjectParentDir)
{
   AutoMapper.ServiceCollectionExtensions.UseStaticRegistration = false;

   var startupAssembly = StartupType.Assembly;
   ...
}
```
was enough for me.

**Comment by jbogard** [comment]
Could you also call ‘Mapper.Reset’ before ‘Services.AddAutoMapper’? If you
wanted to keep static registration.

**Comment by erikaspl** [comment]
@sm-g [SetUp] and [TearDown] will work as well:

```  
    public void Configure(IApplicationBuilder app, IHostingEnvironment env, 
            ILoggerFactory loggerFactory, IApplicationLifetime applicationLifetime)
    {
         .....
         applicationLifetime.ApplicationStopped.Register(() => AutoMapper.Mapper.Reset());
    }
```

```      
        protected TestServer Server;
        [SetUp]
        public void OneTimeBaseSetUp()
        {
             Server = new TestServer(....
        }

        [TearDown]
        public void OneTimeBaseTearDown()
        {
            Server.Dispose();
        }
```

**Comment by DrSensor** [comment]
I also run the same problem and my test are running in [parallel](https://xunit.github.io/docs/running-tests-in-parallel). The only solution that is work for me is by **locking** the process and call `Mapper.Reset()` right before `Mapper.Initialize` is called:
```c#
private static Mutex mutex = new Mutex();

public void ConfigureServices(IServiceCollection services)
{
    mutex.WaitOne();
    Mapper.Reset();
    services.AddAutoMapper(x => x.AddProfile(new QLMapperProfile()), typeof(Startup));
    mutex.ReleaseMutex();
.
.
``` 
I also tried to lock the process and call `Mapper.Reset()` right before `TestServer` is instantiated but it fail.

**Comment by jbogard** [comment]
The static stuff with IServiceCollection just don't seem to mix, at all, so much so that I'm proposing that we remove the static initialization altogether and just to instance-based for `services.AddAutoMapper`.

**Comment by jg11jg** [comment]
I was getting same issue around WebApplicationFactory and AddAutoMapper not working well together.  Directly registering Profiles rather than having AutoMapper do its own discovery worked for me.  Rather than needing to remember to register profiles here as I added them, I did it this way:

            var profileTypes = Assembly.GetExecutingAssembly().GetTypes()
                .Where(x => x.IsSubclassOf(typeof(Profile)));

            services.AddAutoMapper(x => x.AddProfiles(profileTypes), typeof(Startup));
